contracts/example/sources/example.move

```diff
@@ -0,0 +1,132 @@
+module example::connected;
+
+use std::ascii;
+use std::ascii::String;
+use sui::address::from_bytes;
+use sui::coin::Coin;
+use gateway::gateway::{MessageContext, message_context_sender, message_context_target};
+
+// === Errors ===
+
+const EInvalidPayload: u64 = 1;
+
+const EUnauthorizedSender: u64 = 2;
+
+// ENonceMismatch is a fabricated nonce mismatch error code emitted from the on_call function
+// zetaclient should be able to differentiate this error from real withdraw_impl nonce mismatch
+const ENonceMismatch: u64 = 3;
+
+const EInactiveMessageContext: u64 = 4;
+
+const EPackageMismatch: u64 = 5;
+
+// stub for shared objects
+public struct GlobalConfig has key {
+    id: UID,
+    called_count: u64,
+}
+
+public struct Partner has key {
+    id: UID,
+}
+
+public struct Clock has key {
+    id: UID,
+}
+
+public struct Pool<phantom CoinA, phantom CoinB> has key {
+    id: UID,
+}
+
+// share objects
+fun init(ctx: &mut TxContext) {
+    let global_config = GlobalConfig {
+        id: object::new(ctx),
+        called_count: 0,
+    };
+    let pool = Pool<sui::sui::SUI, example::token::TOKEN> {
+        id: object::new(ctx),
+    };
+    let partner = Partner {
+        id: object::new(ctx),
+    };
+    let clock = Clock {
+        id: object::new(ctx),
+    };
+
+    transfer::share_object(global_config);
+    transfer::share_object(pool);
+    transfer::share_object(partner);
+    transfer::share_object(clock);
+}
+
+public entry fun on_call<SOURCE_COIN>(
+    message_context: &MessageContext,
+    in_coins: Coin<SOURCE_COIN>,
+    cetus_config: &mut GlobalConfig,
+    // Note: this pool type is hardcoded as <SUI, TOKEN> and therefore causes type mismatch error in the
+    // fungible token withdrawAndCall test, where the SOURCE_COIN type is FAKE_USDC instead of TOKEN.
+    // Disabling the pool object for now is the easiest solution to allow the E2E tests to go through.
+    // _pool: &mut Pool<SOURCE_COIN, TARGET_COIN>,
+    _cetus_partner: &mut Partner,
+    _clock: &Clock,
+    data: vector<u8>,
+    _ctx: &mut TxContext,
+) {
+    // check if the message is "revert" and revert with faked ENonceMismatch if so
+    if (data == b"revert") {
+        assert!(false, ENonceMismatch);
+    };
+
+    // decode the sender, target package, and receiver from the payload
+    let (authenticated_sender, target_package, receiver) = decode_sender_target_and_receiver(data);
+
+    // check if the sender is the authorized sender
+    let actual_sender = message_context_sender(message_context);
+    assert!(authenticated_sender == actual_sender, EUnauthorizedSender);
+
+    // check if the target package is my own package
```

Comment:
is this checking if its coming from gateway? maybe some clarification on `my own package` would be great